craft_parts/executor/part_handler.py

```diff
@@ -422,14 +538,31 @@ def _run_stage(
         # parameters if overlay contents change.
         overlay_hash = self._compute_layer_hash(all_parts=True)
 
+        migration_partitions_contents: dict[str, MigrationContents] = {
+            p: MigrationContents(files=c.files, directories=c.dirs)
+            for p, c in contents.partitions_contents.items()
+            if p is not DEFAULT_PARTITION
+        }
+
+        default_backstage_files: set[str] = set()
+        default_backstage_directories: set[str] = set()
+
+        default_contents = cast(
+            StagePartitionContents, contents.partitions_contents.get(DEFAULT_PARTITION)
+        )
+        if default_contents:
+            default_backstage_files = default_contents.backstage_files
+            default_backstage_directories = default_contents.backstage_dirs
+
         return states.StageState(
             part_properties=self._part_properties,
             project_options=step_info.project_options,
-            files=contents.files,
-            directories=contents.dirs,
+            partitions_contents=migration_partitions_contents,
+            files=contents.partitions_contents[DEFAULT_PARTITION].files,
+            directories=contents.partitions_contents[DEFAULT_PARTITION].dirs,
             overlay_hash=overlay_hash.hex(),
-            backstage_files=contents.backstage_files,
-            backstage_directories=contents.backstage_dirs,
+            backstage_files=default_backstage_files,
+            backstage_directories=default_backstage_directories,
```

Comment:
The same strategy can be applied here by not declaring `default_backstage_files` and `default_backstage_directories`.